lang2/macro/src/codegen/cross_calling.rs

```diff
@@ -0,0 +1,395 @@
+    }
+
+    fn generate_storage_constructors<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = TokenStream2> + 'a {
+        self.contract
+            .functions
+            .iter()
+            .filter(|function| function.is_constructor())
+            .map(move |function| {
+                let span = function.span();
+                let ident = &function.sig.ident;
+                let attrs = utils::filter_non_ink_attributes(&function.attrs);
+                let fn_args = function.sig.inputs();
+                let arg_idents = function.sig.inputs().map(|fn_arg| &fn_arg.ident);
+                let selector = function.selector().expect("constructors always have selectors");
+                let selector_bytes = selector.as_bytes();
+
+                quote_spanned!(span=>
+                    #( #attrs )*
+                    pub fn #ident(
+                        #( #fn_args ),*
+                    ) -> ink_core::env2::call::CreateBuilder<
+                        Env,
+                        Self,
+                        ink_core::env2::call::state::Sealed,
+                        ink_core::env2::call::state::CodeHashUnassigned,
+                    > {
+                        ink_core::env2::call::CreateParams::<Env, Self>::build(
+                            ink_core::env2::call::Selector::from_bytes([#( #selector_bytes ),*])
+                        )
+                        #(
+                            .push_arg(&#arg_idents)
+                        )*
+                        .seal()
+                    }
+                )
+            })
+    }
+
+    fn generate_storage_messages<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = TokenStream2> + 'a {
+        let storage_ident_lit = self.contract.storage.ident.to_string();
+        self.contract
+            .functions
+            .iter()
+            .filter(|function| function.is_message())
+            .map(move |function| {
+                let span = function.span();
+                let ident = &function.sig.ident;
+                let ident_lit = ident.to_string();
+                let attrs = utils::filter_non_ink_attributes(&function.attrs);
+                let fn_args = function.sig.inputs();
+                let arg_idents = function.sig.inputs().map(|fn_arg| &fn_arg.ident);
+                let output = &function.sig.output;
+                let is_mut = function.sig.is_mut();
+                let call_path = if is_mut {
+                    quote! { ForwardCallMut::call_mut}
+                } else {
+                    quote! { ForwardCall::call }
+                };
+                let receiver = if is_mut {
+                    quote! { &mut self }
+                } else {
+                    quote! { &self }
+                };
+                let failure_msg = match output {
+                    syn::ReturnType::Default => {
+                        format!(
+                            "invokation of {}::{} message was invalid",
```

Comment:
```suggestion
                            "invocation of {}::{} message was invalid",
```

Reply:
I ll never learn this word ... :upside_down_face: